src/libsyntax/parse/parser/stmt.rs

```diff
@@ -41,11 +42,29 @@ impl<'a> Parser<'a> {
         let lo = self.token.span;
 
         Ok(Some(if self.eat_keyword(kw::Let) {
-            Stmt {
-                id: DUMMY_NODE_ID,
-                kind: StmtKind::Local(self.parse_local(attrs.into())?),
-                span: lo.to(self.prev_span),
-            }
+            let local = self.parse_local(attrs.into())?;
+            self.mk_dummy_stmt(lo.to(self.prev_span), local)
+
+        } else if self.is_ident_mut() {
```

Comment:
I'm not sure if this is possible to do this. I think its better to do this kind of error check after we parse statement failed instead in the main parse tree. I had take a dip about this try to do this check after parse `var` or `mut` as a `path` according to the previous parse logic but got lost to find where did the error recovering located. @Centril

Reply:
The reason you need to look-ahead is to prevent e.g. parsing `var();` and similar things the wrong way since that would be a function call.